### PR TITLE
Panzer tangents: ScatterDirichletResiduals Tpetra

### PR DESCRIPTION
**Pull request by reuterb**
@trilinos/panzer 

## Motivation

Continue to update Panzer's tangent capabilities by refreshing Tpetra Dirichlet residual scatters.

## Related Issues

* Closes `put-issue-number-here`

### CODE REVIEW

File: packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp [hunk 1]
```diff
@@ -617,4 +618,252 @@ evaluateFields(typename TRAITS::EvalData workset)
...
+           const int basisIndex = basisIndices(basis);
+
+           // Possible warp divergence for hierarchic
+           if (checkApplyBC)
+             if (!applyBC(cell,basisIndex))
+               continue;
+
+           kokkosScatterTarget(lid,0) = fieldValues(cell,basisIndex).val();
+           for(int i_param=0; i_param<num_params; i_param++)
+             kokkosTangents(i_param)(lid,0) = fieldValues(cell,basis).fastAccessDx(i_param);
```
Comment (reuterb) [inline packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp:846]: @rppawlo This is what I was asking about

File: packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_Tpetra_impl.hpp [hunk 2]
```diff
@@ -332,15 +417,111 @@ preEvaluate(typename TRAITS::PreEvalData d)
...
+
+    // loop over the basis functions (currently they are nodes)
+    for(std::size_t basis=0; basis < offsets.extent(0); basis++) {
+       int offset = offsets(basis);
+       LO lid    = lids(cell,offset);
+       if (lid<0) continue; // not on this processor
+
+       int basisId = basisIds(basis);
+       if (checkApplyBC)
+         if(!applyBC(cell,basisId)) continue;
```
Comment (reuterb) [inline packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_Tpetra_impl.hpp:471]: @rppawlo Also, not sure if we like this branching within the parallel_for. I saw it elsewhere so may not be a concern.

**Comment by reuterb** [comment]
@rppawlo Ready for a look when you get a chance, thanks.

File: packages/panzer/adapters-stk/test/evaluator_tests/tpetra_blocked_scatter_dirichlet_residual.cpp [hunk 3]
```diff
@@ -0,0 +1,1145 @@
...
+   using Thyra_TpetraLinearOp = Thyra::TpetraLinearOp<double, LocalOrdinal, GlobalOrdinal>;
+
+   Teuchos::RCP<panzer::PureBasis> buildBasis(std::size_t worksetSize, const std::string &basisName);
+   void testInitialization(const Teuchos::RCP<Teuchos::ParameterList> &ipb);
+   Teuchos::RCP<panzer_stk::STK_Interface> buildMesh(int elemX, int elemY);
+
+   TEUCHOS_UNIT_TEST(block_assembly, scatter_dirichlet_residual)
+   {
+
+#ifdef HAVE_MPI
```
Comment (rppawlo) [inline packages/panzer/adapters-stk/test/evaluator_tests/tpetra_blocked_scatter_dirichlet_residual.cpp:76]: can probably just remove the ifdefs. MPI is a required TPL dependency for panzer so the else can't happen. I think this was leftover from the early days when it did support a serial build.

**Review by rppawlo — APPROVED** [review]
Looks good to me! Thanks @reuterb ! At some point we should unify all the boilerplate construction of the base panzer objects for all these tests. There's a lot of repeated code that we should just put into a unit test helper library.

File: packages/panzer/adapters-stk/test/evaluator_tests/tpetra_blocked_scatter_dirichlet_residual.cpp [hunk 4]
```diff
@@ -0,0 +1,1145 @@
...
+      panzer::index_t dd_count(0);
+      Teuchos::ArrayRCP<const double> data, dd_data;
+      Teuchos::RCP<const Thyra::ProductVectorBase<double>> f_vec = Teuchos::rcp_dynamic_cast<Thyra::ProductVectorBase<double>>(b_loc->get_f());
+      Teuchos::RCP<const Thyra::ProductVectorBase<double>> dd_vec = Teuchos::rcp_dynamic_cast<Thyra::ProductVectorBase<double>>(b_dd_loc->get_f());
+
+      // check all the residual values. This is kind of crappy test since it simply checks twice the target
+      // value and the target. Its this way because you add two entries across elements.
+
+      Teuchos::rcp_dynamic_cast<const Thyra::SpmdVectorBase<double>>(f_vec->getVectorBlock(0))->getLocalData(Teuchos::ptrFromRef(data));
+      Teuchos::rcp_dynamic_cast<const Thyra::SpmdVectorBase<double>>(dd_vec->getVectorBlock(0))->getLocalData(Teuchos::ptrFromRef(dd_data));
```
[the remark this answers is not in the document]
Comment (reuterb) [reply packages/panzer/adapters-stk/test/evaluator_tests/tpetra_blocked_scatter_dirichlet_residual.cpp:278]: Yeah, I got this pattern from some other tests. I can take a look

**Pull request merged**